File: app/src/main/java/io/github/sheepdestroyer/materialisheep/StoryListViewModel.java
package io.github.sheepdestroyer.materialisheep;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.core.util.Pair;
import android.util.Log;

import io.github.sheepdestroyer.materialisheep.data.Item;
import io.github.sheepdestroyer.materialisheep.data.ItemManager;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;

/**
 * A view model that provides a list of stories.
 */
public class StoryListViewModel extends ViewModel {
    private ItemManager mItemManager;
    private Scheduler mIoThreadScheduler;
    private MutableLiveData<Pair<Item[], Item[]>> mItems; // first = last updated, second = current

    private final io.reactivex.rxjava3.disposables.CompositeDisposable mDisposables = new io.reactivex.rxjava3.disposables.CompositeDisposable();

    /**
     * Injects the dependencies.
     *
     * @param itemManager       The item manager.
     * @param ioThreadScheduler The I/O thread scheduler.
     */
    public void inject(ItemManager itemManager, Scheduler ioThreadScheduler) {
        mItemManager = itemManager;
        mIoThreadScheduler = ioThreadScheduler;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        mDisposables.clear();
    }

    /**
     * Gets the list of stories.
     *
     * @param filter    The filter to apply.
     * @param cacheMode The cache mode.
     * @return The list of stories.
     */
    public LiveData<Pair<Item[], Item[]>> getStories(String filter, @ItemManager.CacheMode int cacheMode) {
        if (mItems == null) {
            mItems = new MutableLiveData<>();
            mDisposables.add(Observable.fromCallable(() -> mItemManager.getStories(filter, cacheMode))
                    .subscribeOn(mIoThreadScheduler)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(items -> setItems(items),
                            t -> Log.e("StoryListViewModel", "Error loading stories", t)));
        }
        return mItems;
    }

    /**
     * Refreshes the list of stories.
     *
     * @param filter    The filter to apply.
     * @param cacheMode The cache mode.
     */
    public void refreshStories(String filter, @ItemManager.CacheMode int cacheMode) {
        if (mItems == null || mItems.getValue() == null) {
            return;
        }
        mDisposables.add(Observable.fromCallable(() -> mItemManager.getStories(filter, cacheMode))
                .subscribeOn(mIoThreadScheduler)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(items -> setItems(items),
                        t -> Log.e("StoryListViewModel", "Error refreshing stories", t)));

    }

    /**
     * Sets the list of items.
     *
     * @param items The list of items.
     */
    void setItems(Item[] items) {
        mItems.setValue(Pair.create(mItems.getValue() != null ? mItems.getValue().second : null, items));
    }
}

File: app/src/main/java/io/github/sheepdestroyer/materialisheep/StoryListViewModel.kt
package io.github.sheepdestroyer.materialisheep

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.github.sheepdestroyer.materialisheep.data.Item
import io.github.sheepdestroyer.materialisheep.data.ItemManager
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * A view model that provides a list of stories.
 *
 * Rewritten to adhere to Modern Android Development standards:
 * - Uses [StateFlow] for UDF.
 * - Uses [viewModelScope] for structured concurrency.
 * - Uses dependency injection for Dispatchers to facilitate testing.
 */
class StoryListViewModel : ViewModel() {

    private lateinit var itemManager: ItemManager
    private var ioDispatcher: CoroutineDispatcher = Dispatchers.IO // Default, but can be injected

    data class StoryState(
        val lastUpdated: List<Item> = emptyList(),
        val current: List<Item> = emptyList(),
        val isLoading: Boolean = false,
        val error: Throwable? = null
    )

    private val _stories = MutableStateFlow(StoryState())
    val stories: StateFlow<StoryState> = _stories.asStateFlow()

    /**
     * Injects the dependencies.
     * Use this method to inject dependencies after ViewModel creation.
     */
    fun inject(itemManager: ItemManager, ioDispatcher: CoroutineDispatcher = Dispatchers.IO) {
        this.itemManager = itemManager
        this.ioDispatcher = ioDispatcher
    }

    /**
     * Gets the list of stories.
     * Triggers a refresh if the current list is empty.
     */
    fun getStories(filter: String?, @ItemManager.CacheMode cacheMode: Int): StateFlow<StoryState> {
        if (_stories.value.current.isEmpty() && !_stories.value.isLoading) {
            refreshStories(filter, cacheMode)
        }
        return stories
    }

    /**
     * Refreshes the list of stories.
     */
    fun refreshStories(filter: String?, @ItemManager.CacheMode cacheMode: Int) {
        if (!::itemManager.isInitialized) {
            Log.e("StoryListViewModel", "ItemManager not injected!")
            _stories.update { it.copy(error = IllegalStateException("ItemManager not injected")) }
            return
        }

        viewModelScope.launch {
            _stories.update { it.copy(isLoading = true, error = null) }
            try {
                val items = withContext(ioDispatcher) {
                    itemManager.getStories(filter, cacheMode)
                }

                _stories.update { state ->
                    state.copy(
                        lastUpdated = state.current,
                        current = items?.toList() ?: emptyList(),
                        isLoading = false
                    )
                }
            } catch (t: Throwable) {
                Log.e("StoryListViewModel", "Error refreshing stories", t)
                _stories.update { it.copy(isLoading = false, error = t) }
            }
        }
    }
}
